resqpy/property/grid_property_collection.py

```diff
@@ -1044,3 +747,403 @@ def write_nexus_collection(self,
+            for i in range(coarsening.coarse_extent_kji[2]):
+                cell_box = coarsening.fine_box_for_coarse(
+                    (k, j, i))  # local box within lgc space of fine cells, for 1 coarse cell
+                # yapf: disable
+                a_coarsened[k, j, i] = np.nansum(a[cell_box[0, 0]:cell_box[1, 0] + 1,
+                                                 cell_box[0, 1]:cell_box[1, 1] + 1,
+                                                 cell_box[0, 2]:cell_box[1, 2] + 1])
+                # yapf: enable
+                if axis is not None:
+                    axis_1 = (axis + 1) % 3
+                    axis_2 = (axis + 2) % 3
+                    # yapf: disable
+                    divisor = ((cell_box[1, axis_1] + 1 - cell_box[0, axis_1]) *
+                               (cell_box[1, axis_2] + 1 - cell_box[0, axis_2]))
+                    # yapf: enable
+                    a_coarsened[k, j, i] = a_coarsened[k, j, i] / float(divisor)
+    return a_coarsened
+
+
+def _coarsening_weighted_mean(coarsening, a, fine_weight, coarse_weight = None, zero_weight_result = np.NaN):
+    a_coarsened = np.empty(tuple(coarsening.coarse_extent_kji))
+    assert a.shape == tuple(coarsening.fine_extent_kji)
+    assert fine_weight.shape == a.shape
+    if coarse_weight is not None:
+        assert coarse_weight.shape == a_coarsened.shape
+    for k in range(coarsening.coarse_extent_kji[0]):
+        for j in range(coarsening.coarse_extent_kji[1]):
+            for i in range(coarsening.coarse_extent_kji[2]):
+                a_coarsened = _coarsening_weighted_mean_singlecell(a_coarsened, a, coarsening, k, j, i, fine_weight,
+                                                                   coarse_weight, zero_weight_result)
+    if coarse_weight is not None:
+        mask = np.logical_or(np.isnan(coarse_weight), coarse_weight == 0.0)
+        a_coarsened = np.where(mask, zero_weight_result, a_coarsened / coarse_weight)
+    return a_coarsened
+
+
+def _coarsening_weighted_mean_singlecell(a_coarsened, a, coarsening, k, j, i, fine_weight, coarse_weight,
+                                         zero_weight_result):
+    cell_box = coarsening.fine_box_for_coarse((k, j, i))  # local box within lgc space of fine cells, for 1 coarse cell
+    a_coarsened[k, j, i] = np.nansum(
+        a[cell_box[0, 0]:cell_box[1, 0] + 1, cell_box[0, 1]:cell_box[1, 1] + 1, cell_box[0, 2]:cell_box[1, 2] + 1] *
+        fine_weight[cell_box[0, 0]:cell_box[1, 0] + 1, cell_box[0, 1]:cell_box[1, 1] + 1,
+                    cell_box[0, 2]:cell_box[1, 2] + 1])
+    if coarse_weight is None:
+        weight = np.nansum(fine_weight[cell_box[0, 0]:cell_box[1, 0] + 1, cell_box[0, 1]:cell_box[1, 1] + 1,
+                                       cell_box[0, 2]:cell_box[1, 2] + 1])
+        if np.isnan(weight) or weight == 0.0:
+            a_coarsened[k, j, i] = zero_weight_result
+        else:
+            a_coarsened[k, j, i] /= weight
+    return a_coarsened
```

Comment:
This function is setting values in the a_coarsened array (in the lines above).  Arrays are passed by reference in python, ie. the calling code will see the changes in the array without having to return the array as well.  As this is a private function, I suggest not returning anything, and removing the re-assignment in the calling code.  This will execute more efficiently.